hibernate-core/src/main/java/org/hibernate/id/uuid/UuidVersion6Strategy.java

```diff
@@ -98,9 +98,9 @@ public UUID generateUuid(SharedSessionContractImplementor session) {
 	private long getSequence(final long currentTimestamp) {
 		lock.lock();
 		try {
-			if ( this.currentTimestamp > currentTimestamp ) {
+			if ( this.currentTimestamp < currentTimestamp ) {
 				this.currentTimestamp = currentTimestamp;
-				clockSequence.set( 0 );
+				clockSequence.updateAndGet( l -> l & 0x1FFFL );
```

Comment:
According to [the specification](https://www.ietf.org/archive/id/draft-peabody-dispatch-new-uuid-format-01.html#name-uuidv6-basic-creation-algor)

>If the state was unavailable (e.g., non-existent or corrupted) or the timestamp is greater than the current timestamp generate a random 14-bit clock sequence value.

So AFAICT, this should rather be

```suggestion
				final byte[] bytes = new byte[2];
				Holder.numberGenerator.nextBytes( bytes );
				final int randomShort = ( ( bytes[0] & 0xFF ) << 8 ) | ( bytes[1] & 0xFF );
				// Use the lower 14 bits
				clockSequence.set( randomShort & 0x2FFF );
```

Also, the `clockSequence` should rather be an `AtomicInteger`. In the code below, you'd have to use `return clockSequence.updateAndGet( i -> ( i + 1) & 0x2FFF);` according to

>If the state was available, but the saved timestamp is less than or equal to the current timestamp, increment the clock sequence value.

Reply:
I was trying to implement (successfully or not, that is another question) according to Method 1 as described in [6.2 Monotonicity and Counters](https://datatracker.ietf.org/doc/html/rfc9562#section-6.2)

It is, of course, possible to generate random counter each time when timestamp changes, but the care should be taken not to generate too large counter that may overflow if too many UUID's generated with same timestamp value. I guess that this is extremely small chance for this, but still non-zero.

Reply:
Change you are suggesting will be better if written simply as:

`clockSequence.updateAndGet( n -> Holder.numberGenerator.nextInt( 1 << 14 ) );`

or maybe with something little bit smaller than 1<<14 , to prevent possible overflow.